<a href="https://colab.research.google.com/github/gustapfp/intro-ml/blob/main/desafio_bix.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [63]:
from sklearn.svm import SVC, LinearSVC
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import pandas as pd
import numpy as np


dataset2020 = pd.read_csv('base_vigente_2020.csv')
dataset_anos_anteriores = pd.read_csv('base_vigente_anos_anteriores.csv', sep=';')

mapa_dataset = {
    'class':'ventilacao'
}

ds_values = {
    'na':0
}
dataset2020.replace(ds_values, inplace=True)
dataset2020 = dataset2020.rename(columns = mapa_dataset)
dataset2020 = dataset2020.replace(['neg', 'pos'], [False, True])
dataset2020

ventilacao  aa_000 ab_000      ac_000  ...  ee_008 ee_009 ef_000 eg_000
0           False      60      0          20  ...       0      0      0      0
1           False      82      0          68  ...       0      0      0      0
2           False   66002      2         212  ...     516      0      0      0
3           False   59816      0        1010  ...  281662   3232      0      0
4           False    1814      0         156  ...      76      0      0      0
...           ...     ...    ...         ...  ...     ...    ...    ...    ...
15995       False   81852      0  2130706432  ...  960024  25566      0      0
15996       False      18      0          52  ...       0      0      0      0
15997       False   79636      0        1670  ...  358934  19548      0      0
15998       False     110      0          36  ...      74      0      0      0
15999       False       8      0           6  ...       0      0      0      0

[16000 rows x 171 columns]

In [66]:
print(dataset2020.loc[:, dataset2020.columns != 'ventilacao'])
print(dataset2020.size)

       aa_000 ab_000      ac_000 ad_000  ...  ee_008 ee_009 ef_000 eg_000
0          60      0          20     12  ...       0      0      0      0
1          82      0          68     40  ...       0      0      0      0
2       66002      2         212    112  ...     516      0      0      0
3       59816      0        1010    936  ...  281662   3232      0      0
4        1814      0         156    140  ...      76      0      0      0
...       ...    ...         ...    ...  ...     ...    ...    ...    ...
15995   81852      0  2130706432    892  ...  960024  25566      0      0
15996      18      0          52     46  ...       0      0      0      0
15997   79636      0        1670   1518  ...  358934  19548      0      0
15998     110      0          36     32  ...      74      0      0      0
15999       8      0           6      4  ...       0      0      0      0

[16000 rows x 170 columns]
2736000


In [64]:
model = LinearSVC()
SEED = 5


x = dataset2020.loc[:, dataset2020.columns != 'ventilacao']
y = dataset2020[['ventilacao']]

np.random.seed(SEED)
x_treino, x_teste, y_treino, y_teste = train_test_split(x, y, test_size=0.25, stratify=y)

model.fit(x_treino, y_treino)
predictions = model.predict(x_teste)

acy_sre_linearSVC = accuracy_score(y_teste, predictions)
print(f"A precisão do modelo LinearSVC foi de: {(acy_sre_linearSVC)*100}")

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


A precisão do modelo LinearSVC foi de: 95.7


/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [4]:
# array = np.arange(10)
# array[0:7]

array([0, 1, 2, 3, 4, 5, 6])

In [5]:
base_predictions = np.ones(len(y_teste), dtype=bool)
base_line_acy_sre = accuracy_score(y_teste, base_predictions)
print(f"A precisão da linha de base foi de: {(base_line_acy_sre)*100}")

A precisão da linha de base foi de: 2.35


In [6]:
base_predictions_zeros = np.zeros(len(y_teste), dtype=bool)
base_line_acy_sre_zeros = accuracy_score(y_teste, base_predictions)
print(f"A precisão da linha de base foi de: {(base_line_acy_sre_zeros)*100}")

A precisão da linha de base foi de: 2.35


In [7]:
model = SVC()

SEED = 5
np.random.seed(SEED)

raw_x_treino, raw_x_teste, y_treino, y_teste = train_test_split(x, y, test_size=0.25, stratify=y)

scaler = StandardScaler()
scaler.fit(raw_x_treino)
x_treino = scaler.transform(raw_x_treino)
x_teste = scaler.transform(raw_x_teste)

model.fit(x_treino, y_treino)
predictionsSVC = model.predict(x_teste)

acy_sre_SVC = accuracy_score(y_teste, predictions)
print(f"A precisão do modelo SVC foi de: {(acy_sre_SVC)*100}")

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


A precisão do modelo SVC foi de: 95.7


In [8]:
print("Comparação entre a precisão dos modelos:")
print('****************************************')
print(f"A precisão do modelo LinearSVC foi de: {(acy_sre_linearSVC)*100}")
print('****************************************')
print(f"A precisão da linha de base 1 foi de: {(base_line_acy_sre)*100}")
print('****************************************')
print(f"A precisão do modelo SVC foi de: {(acy_sre_SVC)*100}")
print('****************************************')
print(f"A precisão da linha de base 0 foi de: {(base_line_acy_sre_zeros)*100}")

Comparação entre a precisão dos modelos:
****************************************
A precisão do modelo LinearSVC foi de: 95.7
****************************************
A precisão da linha de base 1 foi de: 2.35
****************************************
A precisão do modelo SVC foi de: 95.7
****************************************
A precisão da linha de base 0 foi de: 2.35


Se não enviarem nenhum caminhão tem 2.52% de chance de tomar a decisão certa, ou seja tera que pagar maior parte com manutenção corretiva

Se enviar todos os caminhões tem 2.3% de chance de tomar a decisão certa, ou seja tera que pagar maior parte em manutenção preventiva sem defeito

Utilizando o modelo SVC ou LinearSVC tem 95.7% de chance de tomar a decisão correta, ou seja tera que pagar apenas na manutenção preventiva com defeito

In [8]:
# dataset_custo_manutencao = pd.read_csv('custo_manutencao.csv')

In [8]:
# troca = {
#     True: 25,
#     False: 10
# }
# dataset2020['custo_especulado'] = dataset2020.ventilacao.map(troca)
# data

In [9]:
model = LinearSVC()
SEED = 5


x = dataset2020.loc[:, dataset2020.columns != 'ventilacao']
y = dataset2020[['ventilacao']]

np.random.seed(SEED)
x_treino, x_teste, y_treino, y_teste = train_test_split(x, y, test_size=0.25, stratify=y)

model.fit(x_treino, y_treino)
predictions = model.predict(x_teste)

acy_sre_linearSVC = accuracy_score(y_teste, predictions)
print(f"A precisão do modelo LinearSVC foi de: {(acy_sre_linearSVC)*100}")

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


A precisão do modelo LinearSVC foi de: 95.7


/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [77]:
class CalculaCusto:
  def __init__(self, modelo, base_de_dados):
    self.modelo = modelo
    self.base_de_dados = base_de_dados

  def calcula_precisao(self):
    model = self.modelo
    SEED = 5
    np.random.seed(SEED)

    x = self.base_de_dados.loc[:, dataset2020.columns != 'ventilacao']
    y = self.base_de_dados[['ventilacao']]

    x_treino, x_teste, y_treino, y_teste = train_test_split(x, y, test_size=0.25, stratify=y)

    model.fit(x_treino, y_treino)
    prediction = model.predict(x_teste)

    
    precisao = accuracy_score(y_teste, predictions)
    
    return (precisao*100), prediction

  def calcula_custo(self):
    precisao, prediction = self.calcula_precisao()
    taxa_de_erro = 100 - precisao

    tamanho = prediction.size # numero de diagnosticos
    numeros_de_true = np.sum(prediction) # numero de caminhões que tiveram problema no sistema de ventilação (pos)
    numeros_de_false = tamanho - numeros_de_true # nuemro de caminhões que não tiveram problema no sistema de ventilação (neg)

    numero_acerto = int(tamanho*(precisao/100)) # numero de previsões corretas segundo a precisão do modelo
    numero_erro = int(tamanho*(taxa_de_erro/100)) # numero de previsões incorretas segundo a precisão do modelo

    porc_acerto_true = int(numeros_de_true*(precisao/100)) # numero de diagnosticos corretos de problema no sistem de ventilaçaõ acertos
    porc_acerto_false = int(numeros_de_false*(precisao/100)) # numero de diagnosticos corretos de não haver problema no sistem de ventilaçaõ acertos com true

    porc_erros_true = int(numeros_de_true*(taxa_de_erro/100)) # numero de diagnosticos errados de problema no sistem de ventilaçaõ acertos
    porc_erros_false = int(numeros_de_false*(taxa_de_erro/100)) # numero de diagnosticos errados de  não haver problema no sistem de ventilaçaõ acertos
    
    gasto_acerto_true = porc_acerto_true*25 # Gasto com manutenção do sistema de ar com defeito $25
    gasto_acerto_false = porc_acerto_false*0 # Sem gasto com os caminhões com previsões de não ter problema no sistema pois não seriam mandados para a manutenção

    gasto_erros_true = porc_erros_true*500 # Gasto com manutenção corretiva $500
    gasto_erros_false = porc_erros_false*10 # Gasto com vistoria do caminhao pois nao tinham defeito no sistema
    gasto_total_modelo = gasto_acerto_true + gasto_acerto_false + gasto_erros_true + gasto_erros_false
    print(f'O modelo acertou {numero_acerto} diagnoósticos e errou {numero_erro}. Tendo uma precisão de:{precisao}%.')
    print(f'Com isso a IA conseguiu um custo de manutençaõo com os diagnosticos corretos de U${gasto_acerto_true}, com os diagnosticos errados os custos foram de U${gasto_erros_true + gasto_erros_false} totalizando um custo de U${gasto_total_modelo}')
    

In [80]:
custo2020_LinearSVC = CalculaCusto(LinearSVC(), dataset2020)
# previsao_custo2020_LinearSVC = custo2020_LinearSVC.calcula_precisao()
custo2020_LinearSVC.calcula_custo()

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


O modelo acertou 3828 diagnoósticos e errou 171. Tendo uma precisão de:95.7%.
Com isso a IA conseguiu um custo de manutençaõo com os diagnosticos corretos de U$5625, com os diagnosticos errados os custos foram de U$6610 totalizando um custo de U$12235


/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [73]:

precisao, prediction = CalculaCusto(LinearSVC(), dataset2020).calcula_precisao()
taxa_de_erro = 100 - precisao

tamanho = prediction.size
numeros_de_true = np.sum(prediction)
numeros_de_false = tamanho - numeros_de_true
print(f'numeros de true: {numeros_de_true}')
print(f'numero de false: {numeros_de_false}')
print(f'numero de elementos: {tamanho}')
print(f"*****************{dataset2020.size}")
tamanho_acerto = int(tamanho*(precisao/100)) # numero de previsões corretas segundo a precisão do modelo
tamanho_erro = int(tamanho*(taxa_de_erro/100)) # numero de previsões incorretas segundo a precisão do modelo
print(f'numero de acertos do modelo: {tamanho_acerto}')
print(f'numero de erros do modelo: {tamanho_erro}')
print('*************************************************')
print('*************************************************')
porc_acerto_true = int(numeros_de_true*(precisao/100)) # numero de diagnosticos corretos de problema no sistem de ventilaçaõ acertos
porc_acerto_false = int(numeros_de_false*(precisao/100)) # numero de diagnosticos corretos de não haver problema no sistem de ventilaçaõ acertos com true
print(f'numero de acertos com true: {porc_acerto_true}')
print(f'numero de acertos com false: {porc_acerto_false}')
print('*************************************************')
print('*************************************************')
porc_erros_true = int(numeros_de_true*(taxa_de_erro/100)) # numero de diagnosticos errados de problema no sistem de ventilaçaõ acertos
porc_erros_false = int(numeros_de_false*(taxa_de_erro/100)) # numero de diagnosticos errados de  não haver problema no sistem de ventilaçaõ acertos
print(f'numero de erros com true: {porc_erros_true}') 
print(f'numero de erros com false: {porc_erros_false}')
print('*************************************************')
print('*************************************************')
# calculando o custo das predições corretas do modelo
gasto_acerto_true = porc_acerto_true*25
gasto_acerto_false = porc_acerto_false*0

gasto_erros_true = porc_erros_true*500
gasto_erros_false = porc_erros_false*25
gasto_total_modelo = gasto_acerto_true + gasto_acerto_false + gasto_erros_true + gasto_erros_false
print(f'Os gastos do modelo com os diagnosticos corretos são de {gasto_acerto_true}, e os gastos com os diagnosticos errados {gasto_erros_true} e {gasto_erros_false} totalizando {gasto_total_modelo}')

# print(f'previsoes corretas: {previsoes_corretas.size} tamanho acerto:{tamanho_acerto}')

# calculando o custo das predições errada do modelo

# previsoes_erradas = prediction[tamanho_acerto+1:] # array com o numero estimado de previsões erradas

# sr_erros = pd.Series(prediction)

# sr_erros = sr_erros.replace([True, False], [False, True]) # corrige o valor das prediçoes erradas
# previsoes_erradas_corrigidas = sr_erros.to_numpy()
# previsoes_erradas_corrigidas

 
# custo_erro_500 = int((np.sum(previsoes_erradas_corrigidas)) * 500) # Se o caminhão tinha problema e não foi previsto no modelo ele ira ter que fazer uma manutenção corretiva
# custo_erro_10 = int((previsoes_erradas.size - (custo_erro_500/500)) * 10)# Se o caminha tinha não problema e foi previsto que ele havia ele sera levado para oficina e pagaga a taxa da vistoria
# print(custo_erro_500, custo_erro_10)
# custo_erro_total = custo_erro_500 + custo_erro_10
# custo_erro_total



/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


numeros de true: 236
numero de false: 3764
numero de elementos: 4000
*****************2736000
numero de acertos do modelo: 3828
numero de erros do modelo: 171
*************************************************
*************************************************
numero de acertos com true: 225
numero de acertos com false: 3602
*************************************************
*************************************************
numero de erros com true: 10
numero de erros com false: 161
*************************************************
*************************************************
Os gastos do modelo com os diagnosticos corretos são de 5625, e os gastos com os diagnosticos errados 5000 e 4025 totalizando 14650


/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:947: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


In [ ]:
#lixo de debug

# print(prediction.size)
# print(tamanho_acerto)
# print(tamanho_erro)

# troca = {
#     True: False,
#     False: True
# }
# print('*********************')
# df_erros
# .replace(['neg', 'pos'], [False, True])

# print(f'previsoes errada: {previsoes_erradas} tamanho erro:{previsoes_erradas_trocada}')